### ISSUE TITLE
Passenger process stuck at rack_stream.rb:52:in pop

### ISSUE THREAD

**Issue by jordoh** [faye/faye-websocket-ruby#105]
I'm running in to an intermittent issue where passenger processes become stuck waiting for the rack hijack socket to be handled in the EM thread, but the EM reactor is no longer running. Prior to becoming stuck, the passenger process _was_ successfully handling websockets, indicating that the EM reactor thread was running.

Once it becomes stuck, the passenger process continues to receive connections, but they are not serviced, until the passenger process eventually hits my configured limit of 100 connections. Eventually, my configured limit of passenger processes is reached (each with 100 connections not being serviced), resulting in passenger rejecting additional requests until it is restarted.

---

The request that appears to cause this is logged in nginx access logs as:
```
[11/Apr/2017:18:23:42 -0700] "GET / HTTP/1.1" 499 0 "-" "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36" 29.999 - -
```
My understanding is the 499 is nginx saying that a response was not received from the proxy (passenger) - all proxy_*_timeout settings in nginx are configured to 20 seconds (request above shows a total time of 29.999 seconds).

I have a logger output immediately before the following line, which is printed, but logger output immediately after the line is never reached for this request:
```
ws = Faye::WebSocket.new(env, nil)
```

Looking at the passenger process (in this case after it has been stuck for more than 12 hours) I can see the passenger request handling thread's backtrace is:

```
------------------------------------------------------------
# Thread: #<Thread:0x00000002d1f540@/usr/lib/ruby/vendor_ruby/phusion_passenger/utils.rb:94 sleep>(Worker 1), alive = true
------------------------------------------------------------
    [GEM_HOME]/gems/faye-websocket-0.10.5/lib/faye/rack_stream.rb:52:in `pop'
    [GEM_HOME]/gems/faye-websocket-0.10.5/lib/faye/rack_stream.rb:52:in `hijack_rack_socket'
    [GEM_HOME]/gems/faye-websocket-0.10.5/lib/faye/rack_stream.rb:25:in `initialize'
    [GEM_HOME]/gems/faye-websocket-0.10.5/lib/faye/websocket.rb:77:in `new'
    [GEM_HOME]/gems/faye-websocket-0.10.5/lib/faye/websocket.rb:77:in `initialize'
    [PROJECT_ROOT]/lib/touch_up_app.rb:38:in `new'
    [PROJECT_ROOT]/lib/touch_up_app.rb:38:in `handle_request'
    [PROJECT_ROOT]/lib/touch_up_app.rb:13:in `block in call'
    [GEM_HOME]/gems/activesupport-5.0.2/lib/active_support/tagged_logging.rb:69:in `block in tagged'
    [GEM_HOME]/gems/activesupport-5.0.2/lib/active_support/tagged_logging.rb:26:in `tagged'
    [GEM_HOME]/gems/activesupport-5.0.2/lib/active_support/tagged_logging.rb:69:in `tagged'
    [PROJECT_ROOT]/lib/touch_up_app.rb:12:in `call'
    [GEM_HOME]/gems/newrelic_rpm-3.18.0.329/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
    [GEM_HOME]/gems/rack-2.0.1/lib/rack/urlmap.rb:68:in `block in call'
    [GEM_HOME]/gems/rack-2.0.1/lib/rack/urlmap.rb:53:in `each'
    [GEM_HOME]/gems/rack-2.0.1/lib/rack/urlmap.rb:53:in `call'
    [PROJECT_ROOT]/lib/middleware/request_instrumentation.rb:16:in `call'
    [GEM_HOME]/gems/newrelic_rpm-3.18.0.329/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
    [PROJECT_ROOT]/lib/middleware/exception_handling.rb:13:in `call'
    [GEM_HOME]/gems/newrelic_rpm-3.18.0.329/lib/new_relic/agent/instrumentation/middleware_tracing.rb:96:in `call'
    /usr/lib/ruby/vendor_ruby/phusion_passenger/rack/thread_handler_extension.rb:81:in `process_request'
    /usr/lib/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:137:in `accept_and_process_next_request'
    /usr/lib/ruby/vendor_ruby/phusion_passenger/request_handler/thread_handler.rb:98:in `main_loop'
    /usr/lib/ruby/vendor_ruby/phusion_passenger/request_handler.rb:440:in `block (3 levels) in start_threads'
    /usr/lib/ruby/vendor_ruby/phusion_passenger/utils.rb:97:in `block in create_thread_and_abort_on_exception'
```

`passenger-irb` does not show the EM reactor running at this point:
```
>> EventMachine.reactor_running?
=> false
```

I can confirm that `passenger-irb` _does_ show the EM reactor running in a passenger process that is _not_ stuck in this way:

```
>> EventMachine.reactor_running?
=> true
```

and a normal process has the EM reactor thread:

```
------------------------------------------------------------
# Thread: #<Thread:0x000000061ceac0@[GEM_HOME]/gems/faye-websocket-0.10.5/lib/faye/websocket.rb:40 sleep>, alive = true
------------------------------------------------------------
    [GEM_HOME]/gems/eventmachine-1.2.1/lib/eventmachine.rb:194:in `run_machine'
    [GEM_HOME]/gems/eventmachine-1.2.1/lib/eventmachine.rb:194:in `run'
    [GEM_HOME]/gems/faye-websocket-0.10.5/lib/faye/websocket.rb:40:in `block in ensure_reactor_running'
```

but this thread is not included in the threads running in the stuck process.

The passenger and nginx error logs do not contain any output including the PID of the stuck process or the path of the request (the requests have unique paths that include authorization tokens, which I changed to "/" in the nginx access log output above).

---

This is a low-volume application that averages around 30 websocket connections opened per minute and around 150 connections open at any given time. When a client (javascript in web browser) hits a stuck passenger process (prior to 100 connection limit being reached) it handles reconnection, eventually hitting a non-stuck process. However, this takes about 30 seconds each time (nginx timing out the proxy connection) and happens about 1,000 times a day. Passenger still includes the connection in the open session count, eventually leading to the stuck process hitting the 100 connection limit and no longer being used.

**Comment by jcoglan** [comment]
I've never heard of an error like this before, and I was wondering whether you've reduced this to a minimal reproduction case. Nothing inside faye-websocket should be stopping the reactor, so it would be worth finding out what is causing it.

`Faye::WebSocket.new` attempts to [make sure the reactor is running](https://github.com/faye/faye-websocket-ruby/blob/bbef55a2cd7fd52d145cdc73c0d162573b2a83a8/lib/faye/websocket.rb#L69) so something must be stopping that working, or stopping the reactor after that happens.

Are you able to find out what is causing this?

**Comment by jordoh** [comment]
I _believe_ this is due to a well-timed unhandled exception on the reactor thread (possibly in an `em-http-request` callback) stopping the reactor thread before the rack hijack IO block can run. I have added code to verify this, but it has yet to happen it (in ~24 hours) so I can't confirm this is the case.

Assuming that this _is_ the case, would it be reasonable for `hijack_rack_socket` to gracefully handle this edge case? Returning immediately if the reactor is no longer running could let the current request fail, but would allow passenger to continue processing requests. The next websocket request would presumably re-start the reactor.

**Comment by jcoglan** [comment]
@jordoh Apologies for the delay on this, I fell ill for much of May so I've got behind on things.

What you propose sounds like a reasonable idea. Would you like to work up a PR for it?

**Comment by jordoh** [comment]
Hi @jcoglan - sure, I should have a chance to work something up this weekend.

**Comment by jcoglan** [comment]
This is fixed in #106 -- thanks for reporting and fixing it :)
